### PR TITLE
Feature/refactor record classes

### PR DESCRIPTION
**Pull request by scan-lan**
new magnum opus
read ticket [#116](https://zube.io/craig180885-napier/sem-group_project/c/116)

### CODE REVIEW

File: src/main/java/com/napier/sem/Record.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import java.sql.SQLException;
+
+/**
+ * Class representing a database record which could be one of:
+ *   - Country
+ *   - City
+ *   - Capital City
+ * The possible fields consist of:
+ *   code, name, continent, region, country,
+ *   district, capital, population
```
Comment (LukasP23) [inline src/main/java/com/napier/sem/Record.java:14]: Do you want to add recordType here? Not really a field but thought it was worth checking anyway.

File: src/main/java/com/napier/sem/Record.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ */
+public class Record
+{
+    // Private properties
+    private String code;
```
Comment (LukasP23) [inline src/main/java/com/napier/sem/Record.java]: Worth changing this to countryCode?

File: src/test/java/com/napier/sem/DAO_IntegrationTests.java [hunk 3]
```diff
@@ -91,10 +54,10 @@ void allCountriesIn_allAreasMatchFilter()
         String areaFilter = "Africa";
 
         // when
-        ArrayList<Country> countries = dao.allCountriesIn(App.CONTINENT, areaFilter);
+        ArrayList<Record> countries = dao.allCountriesIn(App.CONTINENT, areaFilter);
 
         // then
-        for (Country country: countries) assertEquals(areaFilter, country.getContinent());
+        for (Record record : countries) assertEquals(areaFilter, record.getContinent());
```
Comment (LukasP23) [inline src/test/java/com/napier/sem/DAO_IntegrationTests.java]: @scan-lan  Same issue as we talked about before.

File: src/test/java/com/napier/sem/DAO_IntegrationTests.java [hunk 4]
```diff
@@ -179,10 +142,10 @@ void topNCountriesIn_allAreasMatchFilter()
         String areaName = "Europe";
 
         // when
-        ArrayList<Country> countries = dao.topNCountriesIn(App.CONTINENT, areaName, 10);
+        ArrayList<Record> countries = dao.topNCountriesIn(App.CONTINENT, areaName, 10);
 
         // then
-        for (Country country: countries) assertEquals("Europe", country.getContinent());
+        for (Record record : countries) assertEquals("Europe", record.getContinent());
```
Comment (LukasP23) [inline src/test/java/com/napier/sem/DAO_IntegrationTests.java]: Same again.

File: src/test/java/com/napier/sem/Record_UnitTests.java [hunk 5]
```diff
@@ -4,10 +4,38 @@
...
+
+        // then
+        assertEquals(code, record.getCode());
+        assertEquals(name, record.getName());
+        assertEquals(continent, record.getContinent());
+        assertEquals(region, record.getRegion());
+        assertEquals(population, record.getPopulation());
+        assertEquals(capital, record.getCapital());
+        assertEquals(expectedToStringOutput, record.toString());
+    }
```
Comment (LukasP23) [inline src/test/java/com/napier/sem/Record_UnitTests.java:42]: Haven't asserted expectedRecordType.

File: src/test/java/com/napier/sem/Record_UnitTests.java [hunk 6]
```diff
@@ -17,67 +45,43 @@ public void capitalCity_testGettersAndToString()
...
     @Test
-    public void city_testGettersAndToString()
+    public void Record_testGettersAndToStringForCity()
     {
         // given
         String name = "Wishae";
         String country = "Scotland";
         String district = "North Lanarkshire";
         Integer population = 10000;
+        String expectedRecordType = App.CITY;
```
Comment (LukasP23) [inline src/test/java/com/napier/sem/Record_UnitTests.java:82]: Not a massive issue but these are in a different order than in the method above.

File: src/main/java/com/napier/sem/Record.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+                                "-------------------------------------------------------",
+                        this.name,
+                        this.country,
+                        this.district,
+                        this.population);
+                break;
+        }
+        return recordString;
+    }
+}
```
Comment (Craig180885-napier) [inline src/main/java/com/napier/sem/Record.java:148]: This is a thing of beauty

File: src/main/java/com/napier/sem/Record.java (hunk 1 above)
Comment (scan-lan) [reply src/main/java/com/napier/sem/Record.java:14]: probably not, but we should be mentioning it in the comment i think. i'll update.

File: src/test/java/com/napier/sem/Record_UnitTests.java [hunk 8]
```diff
@@ -4,10 +4,38 @@
```
Comment (LukasP23) [inline src/test/java/com/napier/sem/Record_UnitTests.java:4]: Could we maybe add some quick comments to this file

**Review by LukasP23 — APPROVED** [review]

**Pull request merged**